astor/code_gen.py

```diff
@@ -81,34 +112,108 @@ def set_precedence(value, *nodes):
 set_precedence = precedence_setter()
 
 
+def conditional(*params):
+    """ If the last param evaluates to None, just return
+        an empty tuple
+    """
+    return params if params[-1] is not None else ()
+
+
+def comma_group(items, trailing=False,
+                # Constants
+                set_precedence=set_precedence,
+                Comma=Precedence.Comma, enumerate=enumerate):
+    """ Print a group of 0 or more items, with commas between,
+        and an optional trailing comma.
+    """
+    set_precedence(Comma, *items)
+    yield (((', ', item) if idx else item)
+                for idx, item in enumerate(items))
+    if trailing:
+        yield ','
+
+
+def else_body(node,
+              # Constants
+              If=ast.If, set_precedence=set_precedence,
```

Comment:
The gains do drop off.  I tend to apply them to functions that are called often, and/or reference the attributes in loops.  I'd have to go back and look, but I think I'm getting about an 8% speedup in `to_unparse` (around 5% overall in rtrip) this way.

There were additional gains to be had by adding a proliferation of dispatchers, e.g. adding a `write` method that handled a single item with no for loop, but that was too much stuff to remember, and made the API too big.

I actually went the other way, decomposing `conditional_write` into `condition` and `write`, and coalescing the `write` and `visit` methods into a single `write`, so in one very real sense, these parameter constant micro-optimizations (which I don't mind looking at and thinking about or ignoring, as the case may be) offset the speed loss by making the API more regular and orthogonal and reducing the number of dispatchers.

I also reverted some changes there was a big discussion about last year about whether we should pass a string or a boolean for `async` to node handlers that can handle two things.  Two things decided it for me:  (1) since async is now a keyword, my editor screams at me everywhere it appears, and I think the editor is right -- there's no real reason to use it as a variable name even where you can; and (2) to me this is still more natural (and less code).  Having said that, if anybody still really feels strongly the other way, we could change it back, but they have to pick a different variable name.